capsules/src/sha256.rs

```diff
@@ -0,0 +1,417 @@
+            if self.handle.is_none() {
+                Err((ErrorCode::FAIL, data.take()))
+            } else {
+                let len = data.len();
+                self.handle.map(|handle| {
+                    self.deferred_caller.set(*handle);
+                    self.input_data.set(data);
+                    self.compute_sha256();
+                });
+                Ok(len)
+            }
+        }
+    }
+
+    fn clear_data(&self) {
+        let _ = self.initialize();
+    }
+}
+
+impl<'a> DigestHash<'a, 32> for Sha256Software<'a> {
+    fn run(
+        &'a self,
+        digest: &'static mut [u8; 32],
+    ) -> Result<(), (ErrorCode, &'static mut [u8; 32])> {
+        if self.busy() {
+            Err((ErrorCode::BUSY, digest))
+        } else {
+            self.state.set(State::Hash);
+            if self.handle.is_none() {
+                Err((ErrorCode::FAIL, digest))
+            } else {
+                self.handle.map(|handle| {
+                    self.complete_sha256();
+                    for i in 0..8 {
+                        let val = self.hash_values.get()[i];
+                        digest[4 * i + 3] = (val >> 0 & 0xff) as u8;
+                        digest[4 * i + 2] = (val >> 8 & 0xff) as u8;
+                        digest[4 * i + 1] = (val >> 16 & 0xff) as u8;
+                        digest[4 * i + 0] = (val >> 24 & 0xff) as u8;
+                    }
+                    self.output_data.set(Some(digest));
+                    self.deferred_caller.set(*handle);
+                });
+                Ok(())
+            }
+        }
+    }
+}
+
+impl<'a> DigestVerify<'a, 32> for Sha256Software<'a> {
+    fn verify(
+        &'a self,
+        compare: &'static mut [u8; 32],
+    ) -> Result<(), (ErrorCode, &'static mut [u8; 32])> {
+        if self.busy() {
+            Err((ErrorCode::BUSY, compare))
+        } else {
+            self.state.set(State::Verify);
+            if self.handle.is_none() {
+                Err((ErrorCode::FAIL, compare))
+            } else {
+                self.handle.map(|handle| {
+                    self.complete_sha256();
+                    self.output_data.set(Some(compare));
+                    self.deferred_caller.set(*handle);
+                });
+                Ok(())
+            }
+        }
+    }
+}
+
+impl<'a> Digest<'a, 32> for Sha256Software<'a> {
+    fn set_client(&'a self, client: &'a dyn Client<'a, 32>) {
+        self.client.set(client);
+    }
+}
+
+impl<'a> DynamicDeferredCallClient for Sha256Software<'a> {
+    fn call(&self, _handle: DeferredCallHandle) {
+        let prior = self.state.get();
+        self.state.set(State::Idle);
+        match prior {
+            State::Idle => {}
+            State::Verify => {
+                // Do the verification here so we don't have to store
+                // the result across the callback.
+                let output = self.output_data.replace(None).unwrap();
```

Comment:
Can we remove the `unwrap()`

Reply:
I'm of the opinion that there should be a fail fast here. If the unwrap isn't possible, you can't invoke the callback. If you can't invoke the callback, the state machine is wedged. This should panic, as it implies an internal error that is going to hang the system.